liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,125 @@
+                           || name `S.member` ctors
+                           || isPrefixOfSym anfPrefix name)
+
+flatten :: Expr -> [Expr]
+flatten (EApp f e) = flatten f ++ [e]
+flatten e = [e]
+
+groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
+groupUnifiableEqualities = mapMaybe (keep2 . snd) . M.groupList 
+    where -- We perform only 2-way unification
+          keep2 [e1, e2] = Just (e1, e2)
+          keep2 _        = Nothing
+
+getEqualities :: Expr -> [(Expr, Expr)]
+getEqualities (PAtom Eq e1 e2) = [(e1, e2)]
+getEqualities (PAnd eqs)       = concatMap getEqualities eqs
+getEqualities _                = []
+
+
+-- +-----------------------------------------------------+
+-- | AcyclicRewrites: collection of rewrites that cannot |
+-- | cause an infinite loop                              |
+-- +-----------------------------------------------------+
+-- This could be extracted as a separate module
+
+-- | A collection of rewrites that cannot cause an infinite loop
+newtype AcyclicRewrites = AR (M.HashMap Symbol Expr)
+
+-- We can think of the map a directed graph where every symbol is a vertex and
+-- there is an edge v₁ → v₂ if in the expression that v₁ is rewritten to v₂ is a
+-- free variable.  To guarantee that there arent any infinite loops in the
+-- rewrite procedure that will be executed by PLE, we need to esnure that there
+-- aren't any cyclic rewrites, this is ensured by adding the invariant to our
+-- graph of beeing a DAG.
+
+-- | Takes a list of rewrites and returns a list of rewrites and filters out any
+-- rewrites that would cause an infinite loop, the procedure is order biased as
+-- rewrites earlier in the list take precedence.
+unloop :: [(Symbol, Expr)] -> LocalRewrites
+unloop = LocalRewrites . toRewrites . foldl' doInsert empty 
+    where doInsert ar (s, e) = ar `fromMaybe` insert ar s e 
+
+-- | Get the "raw" list of rewrites
+toRewrites :: AcyclicRewrites -> M.HashMap Symbol Expr
+toRewrites (AR m) = m
+
+-- | Checks if there is a path from s1 to s2 O(Σ(e ∈ m) |exprSymbolSet e|) [O(m)
+-- where m is the number of edges]
+existsPath :: AcyclicRewrites -> Symbol -> Symbol -> Bool
+existsPath (AR m) s1' s2 = go s1'
+  where
+    -- Since m is a DAG, we can use DFS to check if there is a path from s1 to
+    -- s2 without worrying about infinite loops
+    go s1 | s1 == s2 = True
+    go s1 | Just e <- M.lookup s1 m
+          = any go $ S.toList $ exprSymbolsSet e
+    go _ = False
+
+-- | Constructs an empty set of rewrites
+empty :: AcyclicRewrites
+empty = AR M.empty
+
+-- | Inserts the rewrite if it wont't cause an infinite loop
+insert :: AcyclicRewrites -> Symbol -> Expr -> Maybe AcyclicRewrites
+insert ar s e | not $ any (existsPath ar s) $ S.toList $ exprSymbolsSet e
+              , not $ s `S.member` exprSymbolsSet e
+              -- There are two ways to break the DAG invariant
+              -- 1. If the rewrite is closing a loop
+              -- 2. If the rewrite by itself is a cycle
```

Comment:
`existsPath` detects both conditions. Is it necessary to test for (2) separately?

Reply:
exists path is only looking for paths that are already in the path if we have the isolated node x and some new equation x = id x it wont discover the loop

Reply:
My bad.